File: go/datastore/tablet_writer.go
package datastore

import (
	"bytes"
	"code.google.com/p/snappy-go/snappy"
	"encoding/binary"
	"io"
	"log"
)

const (
	tabletMagic    uint32 = 0x0b501e7e
	metaIndexMagic uint32 = 0x0ea7da7a
	dataIndexMagic uint32 = 0xda7aba5e
)

type BlockHandle struct {
	offset uint64
	length uint64
}

type IndexRecord struct {
	offset uint64
	length uint32
	name   []byte
}

type TabletOptions struct {
	BlockSize          uint32
	BlockCompression   BlockCompressionType
	KeyRestartInterval uint
}

func WriteTablet(w io.Writer, kvs Iterator, opts *TabletOptions) {
	headLen := uint64(writeHeader(w, opts))
	dataBlocks := writeDataBlocks(w, kvs, headLen, opts)
	metaIndexLen := writeIndex(w, metaIndexMagic, nil)

	dataIndexLen := writeIndex(w, dataIndexMagic, dataBlocks)

	lastBlock := dataBlocks[len(dataBlocks)-1]
	metaPos := lastBlock.offset + uint64(lastBlock.length)

	metaIndexHandle := BlockHandle{metaPos, metaIndexLen}
	dataIndexHandle := BlockHandle{metaPos + metaIndexLen, dataIndexLen}

	writeFooter(w, metaIndexHandle, dataIndexHandle)
}

func writeDataBlocks(w io.Writer, kvs Iterator, pos uint64, opts *TabletOptions) []*IndexRecord {
	index := make([]*IndexRecord, 0)
	bw := NewBlockWriter(opts)

	flushBlock := func() {
		rec := writeBlock(w, pos, bw, opts)
		bw.Reset()

		if rec == nil {
			return
		}

		index = append(index, rec)
		pos += uint64(rec.length)
	}

	var prevKey []byte
	for kvs.Next() {
		key := kvs.Key()
		if prevKey != nil && bytes.Compare(prevKey, key) >= 0 {
			log.Printf("writing non-increasing keys: %s -> %s",
				prevKey, key)
		}

		bw.Append(kvs.Key(), kvs.Value())

		if bw.Size() > opts.BlockSize {
			flushBlock()
		}
	}

	flushBlock()

	kvs.Close()
	return index
}

func writeBlock(w io.Writer, pos uint64, bw *BlockWriter, opts *TabletOptions) *IndexRecord {
	firstKey, block := bw.Finish()
	if firstKey == nil {
		// empty block
		return nil
	}

	comp, blockType, _ := compress(opts, block)

	var checksum uint32 = 0
	var length uint32 = uint32(len(comp))

	a := writeUint(w, uint(checksum))
	b := writeUint(w, uint(blockType))
	c := writeUint(w, uint(length))

	length += uint32(a + b + c)

	w.Write(comp)

	return &IndexRecord{pos, length, firstKey}
}

func compress(opts *TabletOptions, input []byte) ([]byte, BlockCompressionType, error) {
	var buf bytes.Buffer

	switch opts.BlockCompression {
	case Snappy:
		comp, err := snappy.Encode(nil, input)
		if len(comp) > len(input) {
			// no gains, write an uncompressed block
			buf.Write(input)
			return buf.Bytes(), None, err
		} else {
			buf.Write(comp)
			return buf.Bytes(), Snappy, err
		}
	default:
		return input, None, nil
	}
}

func writeKv(w io.Writer, key []byte, value []byte) uint32 {
	keyCount := writeRaw(w, key)
	valCount := writeRaw(w, value)

	return keyCount + valCount
}

func writeIndex(w io.Writer, magic uint32, recs []*IndexRecord) uint64 {
	binary.Write(w, binary.BigEndian, magic)

	var n uint64
	for _, rec := range recs {
		n += uint64(writeUint(w, uint(rec.offset)))
		n += uint64(writeUint(w, uint(rec.length)))
		n += uint64(writeRaw(w, rec.name))
	}

	// include 4 bytes for the magic number
	return n + 4
}

func writeBinary(w io.Writer, data interface{}) (uint32, error) {
	n := uint32(binary.Size(data))
	err := binary.Write(w, binary.BigEndian, data)
	return n, err
}

func writeHeader(w io.Writer, opts *TabletOptions) uint32 {
	h := Header{tabletMagic, BlockEncodingType(PrefixCompressed), 0, 0, 0}
	n, _ := writeBinary(w, &h)
	return n
}

func writeFooter(w io.Writer, meta BlockHandle, data BlockHandle) {
	writeUint64(w, meta.offset)
	writeUint64(w, meta.length)
	writeUint64(w, data.offset)
	writeUint64(w, data.length)
	binary.Write(w, binary.BigEndian, tabletMagic)
}

File: java/src/main/java/com/thefactory/datastore/TabletWriter.java
package com.thefactory.datastore;

import org.xerial.snappy.Snappy;

import java.io.ByteArrayOutputStream;
import java.io.DataOutput;
import java.io.DataOutputStream;
import java.io.IOException;
import java.nio.ByteBuffer;
import java.nio.channels.WritableByteChannel;
import java.util.Deque;
import java.util.Iterator;
import java.util.LinkedList;
import java.util.zip.CRC32;

public class TabletWriter {
    TabletWriterOptions opts;

    public TabletWriter(TabletWriterOptions opts) {
        this.opts = opts;
    }

    public void writeTablet(WritableByteChannel out, Iterator<KV> kvs) throws IOException {
        ByteArrayOutputStream buf = new ByteArrayOutputStream();

        int headLen = flush(out, writeHeader(buf, opts));

        Deque<IndexRecord> dataBlocks = writeDataBlocks(out, kvs, headLen, opts);

        int metaIndexLen = flush(out, writeIndex(buf, TabletConstants.META_INDEX_MAGIC, new LinkedList<IndexRecord>()));
        int dataIndexLen = flush(out, writeIndex(buf, TabletConstants.DATA_INDEX_MAGIC, dataBlocks));

        IndexRecord lastBlock = dataBlocks.getLast();
        long metaPos = lastBlock.offset + lastBlock.length;

        BlockHandle metaIndexHandle = new BlockHandle(metaPos, metaIndexLen);
        BlockHandle dataIndexHandle = new BlockHandle(metaPos + metaIndexLen, dataIndexLen);

        flush(out, writeFooter(buf, metaIndexHandle, dataIndexHandle));
    }

    /* flush the data in buf to out, resetting buf and returning bytes written */
    private int flush(WritableByteChannel out, ByteArrayOutputStream buf) throws IOException {
        out.write(ByteBuffer.wrap(buf.toByteArray()));

        int count = buf.size();
        buf.reset();
        return count;
    }

    private ByteArrayOutputStream writeHeader(ByteArrayOutputStream out, TabletWriterOptions opts) throws IOException {
        DataOutputStream dos = new DataOutputStream(out);

        dos.writeInt(TabletConstants.TABLET_MAGIC);

        // 0x01: prefix-compressed blocks
        // 0x000000: reserved for future use
        dos.writeInt(0x01000000);

        return out;
    }

    private Deque<IndexRecord> writeDataBlocks(WritableByteChannel out, Iterator<KV> kvs, int pos, TabletWriterOptions opts) throws IOException {
        Deque<IndexRecord> index = new LinkedList<IndexRecord>();
        BlockWriter bw = new BlockWriter(opts);

        while (kvs.hasNext()) {
            KV kv = kvs.next();
            bw.append(kv.getKeyBytes(), kv.getValueBytes());

            if (bw.size() > opts.blockSize) {
                index.add(flushBlock(out, pos, bw, opts));
                pos += index.getLast().length;
            }
        }

        if (bw.getFirstKey() != null) {
            index.add(flushBlock(out, pos, bw, opts));
            pos += index.getLast().length;
        }

        return index;
    }

    private IndexRecord flushBlock(WritableByteChannel out, int pos, BlockWriter bw, TabletWriterOptions opts) throws IOException {
        byte[] firstKey = bw.getFirstKey();
        byte[] data = bw.finish();
        byte blockFlags = 0x00; // uncompressed block

        if (opts.useCompression) {
            byte[] compressed = Snappy.compress(data);
            if (compressed.length < data.length) {
                blockFlags = 0x01;
                data = compressed;
            }
        }

        // write the block envelope: checksum, flags, and length
        ByteArrayOutputStream env = new ByteArrayOutputStream(10 + firstKey.length);
        DataOutput dos = new DataOutputStream(env);

        // envelope is: checksum, block flags, length
        Msgpack.writeUint(dos, getChecksum(data));
        Msgpack.writeUint(dos, blockFlags);

        Msgpack.writeUint(dos, data.length);

        out.write(ByteBuffer.wrap(env.toByteArray()));
        out.write(ByteBuffer.wrap(data));

        bw.reset();

        return new IndexRecord(pos, env.size() + data.length, firstKey);
    }

    private long getChecksum(byte[] data) {
        CRC32 crc32 = new CRC32();
        crc32.update(data);
        return crc32.getValue();
    }

    private ByteArrayOutputStream writeIndex(ByteArrayOutputStream out, int magic, Deque<IndexRecord> recs) throws IOException {
        DataOutput dos = new DataOutputStream(out);
        dos.writeInt(magic);

        for (IndexRecord rec: recs) {
            Msgpack.writeUint(dos, rec.offset);
            Msgpack.writeUint(dos, rec.length);
            Msgpack.writeRaw(dos, rec.name);
        }

        return out;
    }

    private ByteArrayOutputStream writeFooter(ByteArrayOutputStream out, BlockHandle metaIndexHandle, BlockHandle dataIndexHandle) throws IOException {
        DataOutput dos = new DataOutputStream(out);
        Msgpack.writeUint64(dos, metaIndexHandle.offset);
        Msgpack.writeUint64(dos, metaIndexHandle.length);
        Msgpack.writeUint64(dos, dataIndexHandle.offset);
        Msgpack.writeUint64(dos, dataIndexHandle.length);
        dos.writeInt(TabletConstants.TABLET_MAGIC);

        return out;
    }

    public class BlockHandle {
        public BlockHandle(long offset, long length) {
            this.offset = offset;
            this.length = length;
        }

        long offset;
        long length;
    }

    public class IndexRecord {
        public IndexRecord(long offset, int length, byte[] name) {
            this.offset = offset;
            this.length = length;
            this.name = name;
        }

        long offset;
        int length;
        byte[] name;
    }
}
